src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6271,229 +6271,205 @@ private TypeSyntax ParsePointerTypeMods(TypeSyntax type)
+                        return this.ParseIfStatement();
+                    case SyntaxKind.ElseKeyword:
+                        // Including 'else' keyword to handle 'else without if' error cases 
+                        return this.ParseMisplacedElse();
+                    case SyntaxKind.LockKeyword:
+                        return this.ParseLockStatement();
+                    case SyntaxKind.ReturnKeyword:
+                        return this.ParseReturnStatement();
+                    case SyntaxKind.SwitchKeyword:
+                        return this.ParseSwitchStatement();
+                    case SyntaxKind.ThrowKeyword:
+                        return this.ParseThrowStatement();
+                    case SyntaxKind.UnsafeKeyword:
+                        result = TryParseStatementStartingWithUnsafe();
+                        if (result != null)
+                            return result;
+                        break;
+                    case SyntaxKind.UsingKeyword:
+                        return ParseStatementStartingWithUsing();
+                    case SyntaxKind.WhileKeyword:
+                        return this.ParseWhileStatement();
+                    case SyntaxKind.OpenBraceToken:
+                        return this.ParseBlock();
+                    case SyntaxKind.SemicolonToken:
+                        return _syntaxFactory.EmptyStatement(this.EatToken());
+                    case SyntaxKind.IdentifierToken:
+                        result = TryParseStatementStartingWithIdentifier(isGlobalScriptLevel);
+                        if (result != null)
+                            return result;
+                        break;
                 }
 
-                // We could not successfully parse the statement as a non-declaration. Try to parse
-                // it as either a declaration or as an "await X();" statement that is in a non-async
-                // method. 
-
-                return TryParsePossibleDeclarationOrBadAwaitStatement();
+                return ParseStatementCoreRest(isGlobalScriptLevel, ref resetPointBeforeStatement);
             }
             finally
             {
                 _recursionDepth--;
+                this.Release(ref resetPointBeforeStatement);
+            }
+
+            bool canReuseStatement(bool isGlobalScriptLevel)
+            {
+                // Can only reuse a global-script or regular statement if we're in the same
+                // global-script context we were originally in.
+                return this.IsIncrementalAndFactoryContextMatches &&
+                       this.CurrentNode is Syntax.StatementSyntax &&
+                       isGlobalScriptLevel == (this.CurrentNode.Parent is Syntax.GlobalStatementSyntax);
             }
         }
 
-        private StatementSyntax TryParsePossibleDeclarationOrBadAwaitStatement()
+        private StatementSyntax ParseStatementCoreRest(bool isGlobalScriptLevel, ref ResetPoint resetPointBeforeStatement)
```

Comment:
I'm wondering why some methods have a Try prefix and others do not. It seems like some of the methods that do not have the Try prefix can still fail to parse (indicated by returning null).

Reply:
yes.  i'd like to make all the ones that can return null be 'try' methods. but we're not there yet.  it would also be good to be able to make this all `#nullable enabl`  but that will be a big item